cloud/pkg/router/listener/message.go

```diff
@@ -0,0 +1,109 @@
+package listener
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"k8s.io/klog/v2"
+
+	beehiveContext "github.com/kubeedge/beehive/pkg/core/context"
+	"github.com/kubeedge/beehive/pkg/core/model"
+)
+
+var MessageHandlerInstance = &MessageHandler{}
+
+type MessageHandler struct {
+	handlers         sync.Map
+	callbackHandlers sync.Map
+}
+
+func (mh *MessageHandler) AddListener(key interface{}, han Handle) {
+	resource, ok := key.(string)
+	if !ok {
+		return
+	}
+	mh.handlers.Store(resource, han)
+}
+func (mh *MessageHandler) RemoveListener(key interface{}) {
+	resource, ok := key.(string)
+	if !ok {
+		return
+	}
+	mh.handlers.Delete(resource)
+}
+
+func (mh *MessageHandler) getHandler(source string, resource string) (Handle, error) {
+	rs := strings.Split(resource, "/")
+	//todo： check eventbus to router
+	if len(rs) >= 2 && (rs[0] == "rule-endpoint" || rs[0] == "rule") {
+		resource = rs[0]
+	}
+	key := fmt.Sprintf("%s/%s", source, resource)
+	v, exist := mh.handlers.Load(key)
+	if !exist {
+		return nil, errors.New("No handler for message")
+	}
+	handle, ok := v.(Handle)
+	if !ok {
+		return nil, fmt.Errorf("Handler invalid, key is %s", key)
+	}
+	return handle, nil
+}
+
+func Process(module string) {
```

Comment:
I only see one place calls it, is it necessary to make it as a function?

Reply:
yes. it can make the logic clear